plugins/rbac-backend/src/role-manager/ancestor-search-memo.ts

```diff
@@ -0,0 +1,119 @@
+import { TokenManager } from '@backstage/backend-common';
+import { CatalogApi } from '@backstage/catalog-client';
+import { Entity } from '@backstage/catalog-model';
+
+import { alg, Graph } from '@dagrejs/graphlib';
+import { Logger } from 'winston';
+
+// AncestorSearchMemo - should be used to build group hierarchy graph for User entity reference.
+// It supports search group entity reference link in the graph.
+// Also AncestorSearchMemo supports detection cycle dependencies between groups in the graph.
+//
+export class AncestorSearchMemo {
+  private graph: Graph;
+
+  private tokenManager: TokenManager;
+  private catalogApi: CatalogApi;
+
+  private userEntityRef: string;
+
+  private allGroups: Entity[];
+
+  constructor(
+    userEntityRef: string,
+    tokenManager: TokenManager,
+    catalogApi: CatalogApi,
+  ) {
+    this.graph = new Graph({ directed: true });
+    this.userEntityRef = userEntityRef.split('/')[1];
```

Comment:
This field actually is not userEntityRef any more (after split with `/`). I think it should have a name userName or something like that.